java/ql/src/semmle/code/java/security/JWT.qll

```diff
@@ -0,0 +1,173 @@
+ */
+class JwtParserWithSigningKeyExpr extends Expr {
+  MethodAccess signingMa;
+
+  JwtParserWithSigningKeyExpr() { isSigningKeySetter(this, signingMa) }
+
+  /** Gets the method access that sets the signing key for this parser. */
+  MethodAccess getSigningMethodAccess() { result = signingMa }
+}
+
+/**
+ * The qualifier of an insecure parsing method.
+ * That is, either the qualifier of a call to a `parse(token)`, `parseClaimsJwt(token)` or `parsePlaintextJwt(token)` method or
+ * the qualifier of a call to a `parse(token, handler)` method where the `handler` is considered insecure.
+ */
+class JwtParserWithInsecureParseSink extends DataFlow::Node {
+  MethodAccess insecureParseMa;
+
+  JwtParserWithInsecureParseSink() {
+    insecureParseMa.getQualifier() = this.asExpr() and
+    (
+      sinkNode(this, "jwt-insecure-parse")
+      or
+      sinkNode(this, "jwt-insecure-parse-handler") and
+      isInsecureJwtHandler(insecureParseMa.getArgument(1))
+    )
+  }
+
+  /** Gets the method access that does the insecure parsing. */
+  MethodAccess getParseMethodAccess() { result = insecureParseMa }
+}
+
+/**
+ * Holds if `signingMa` directly or indirectly sets a signing key for `expr`, which is a `JwtParser`.
+ * The `setSigningKey` and `setSigningKeyResolver` methods set a signing key for a `JwtParser`.
+ *
+ * Directly means code like this (the signing key is set directly on a `JwtParser`):
+ * ```java
+ * Jwts.parser().setSigningKey(key).parse(token);
+ * ```
+ *
+ * Indirectly means code like this (the signing key is set on a `JwtParserBuilder` indirectly setting the key of `JwtParser` that is created by the call to `build`):
+ * ```java
+ * Jwts.parserBuilder().setSigningKey(key).build().parse(token);
+ * ```
+ */
+private predicate isSigningKeySetter(Expr expr, MethodAccess signingMa) {
+  any(SigningToInsecureMethodAccessDataFlow s)
+      .hasFlow(DataFlow::exprNode(signingMa), DataFlow::exprNode(expr))
+}
+
+/**
+ * Holds if `parseHandlerExpr` is an insecure `JwtHandler`.
+ * That is, it overrides a method from `JwtHandlerOnJwtMethod` and the override is not defined on `JwtHandlerAdapter`.
+ * `JwtHandlerAdapter`'s overrides are safe since they always throw an exception.
+ */
+private predicate isInsecureJwtHandler(Expr parseHandlerExpr) {
+  exists(RefType t |
+    parseHandlerExpr.getType() = t and
+    t.getASourceSupertype*() instanceof TypeJwtHandler and
+    exists(Method m |
+      m = t.getAMethod() and
+      m.getASourceOverriddenMethod+() instanceof JwtHandlerOnJwtMethod and
+      not m.getSourceDeclaration() instanceof JwtHandlerAdapterOnJwtMethod
+    )
+  )
+}
+
+/** CSV source models representing methods that assign signing keys to a JWT parser. */
+private class SigningKeySourceModel extends SourceModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        "io.jsonwebtoken;JwtParser;true;setSigningKey;;;ReturnValue;jwt-signing-key",
```

Comment:
The jjwt library also has `DefaultJwtParser` and `DefaultJwtParserBuilder`.
A user could dirrectly instantiate them, I should have added tests for this case to my PR :D 

I don't think that many are using the `DefaultFoo` classes, but having them here for completeness would be good I think.
If I had added tests, the original query should have flagged them.

Reply:
You actually have that in one of your tests :) https://github.com/github/codeql/pull/5911/files#diff-f2a84795838df366085dccf0b4db85125c5b9fe468dee455341454f6bd3cc639L26

If I'm not mistaken, the third column of the CSV (`true`) should take that into account. According to the `ExternalFlow` QLDocs:

> 3. The `subtypes` is a boolean that indicates whether to jump to an arbitrary subtype of that type.

To make sure, I added an additional test that checks the `DefaultJwtParserBuilder` case in 34a55e7.

Thanks for pointing this out!

Reply:
> You actually have that in one of your tests :) https://github.com/github/codeql/pull/5911/files#diff-f2a84795838df366085dccf0b4db85125c5b9fe468dee455341454f6bd3cc639L26

Missed that while skimming it :facepalm: 
> If I'm not mistaken, the third column of the CSV (`true`) should take that into account. According to the `ExternalFlow` QLDocs:
> 
> > 1. The `subtypes` is a boolean that indicates whether to jump to an arbitrary subtype of that type.
> 

Thanks for the explanation, I've not really worked with CSV modeling before :)